Use documentation comments from inherited properties when @inheritDoc is present

sjbarag opened the pull request:
The JSDoc `@ineheritDoc` [tag](http://usejsdoc.org/tags-inheritdoc.html) "indicates that a symbol should inherit its documentation from its parent class".  In the case of a TypeScript file, this also includes implemented interfaces and parent interfaces.

With this change, a class method or property (or an interface property) with the `@inheritDoc` tag in its JSDoc comment will automatically use the comments from its nearest ancestor that has no `@inheritDoc` tag. To prevent breaking backwards compatibility, `Symbol.getDocumentationComment` now accepts an optional `TypeChecker` instance to support this feature.

fixes #8912

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=18804) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=18804) before we can accept your contribution.<br/><br/>:x: sjbarag [sign now](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=18804)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/TypeScript?pullRequest=18804) it.</sub>

mhegazy commented:
@andy-ms can you please review this change.

sjbarag commented:
For posterity: rebased on fcb48dd9 and re-ran `gulp baseline-accept` to accept changes to `api/tsserverlibrary.d.ts` and `api/typescript.d.ts`.  I haven't changed the diff otherwise though 😄

andy-ms commented on src/services/services.ts:
```diff
@@ -471,9 +481,13 @@ namespace ts {
             if (this.documentationComment === undefined) {
                 this.documentationComment = this.declaration ? JsDoc.getJsDocCommentsFromDeclarations([this.declaration]) : [];
+
+                if (this.declaration && this.documentationComment.length === 0 && hasJSDocInheritDocTag(this) && typeChecker) {
```
This might whole block might be neater as `if (this.declaration) { ... } else { this.documentationComment = []; }`

andy-ms commented on src/services/services.ts:
```diff
@@ -488,6 +502,63 @@ namespace ts {
+     * @returns `true` if `symbol` has a JSDoc "inheritDoc" tag on it, otherwise `false`.
+     */
+    function hasJSDocInheritDocTag(symbol: Signature | Symbol) {
+        return !!find(symbol.getJsDocTags(), tag => tag.name === "inheritDoc");
```
`symbol.getJsDocTags().some(...)`

andy-ms commented on src/services/jsDoc.ts:
```diff
@@ -70,7 +71,7 @@ namespace ts.JsDoc {
         forEachUnique(declarations, declaration => {
             for (const tag of getJSDocTags(declaration)) {
-                if (tag.kind === SyntaxKind.JSDocTag) {
+                if (tag.kind === SyntaxKind.JSDocTag || tag.kind === SyntaxKind.JSDocInheritDocTag) {
```
Don't think you need to modify this function for what you want; use `ts.getJSDocTags` (from `compiler/utilities.ts`) instead, which returns tags of all kinds.

andy-ms commented on src/services/services.ts:
```diff
@@ -488,6 +502,63 @@ namespace ts {
+    function findInheritedJSDocComments(declaration: Declaration, propertyName: string, typeChecker: TypeChecker): SymbolDisplayPart[] {
+        let documentationComment: SymbolDisplayPart[] = [];
+
+        if (isClassDeclaration(declaration.parent) || isInterfaceDeclaration(declaration.parent)) {
```
Prefer an early return. Also, use `isClassLike`.

andy-ms commented on src/services/services.ts:
```diff
@@ -488,6 +502,63 @@ namespace ts {
+        let documentationComment: SymbolDisplayPart[] = [];
+
+        if (isClassDeclaration(declaration.parent) || isInterfaceDeclaration(declaration.parent)) {
+            const container: ClassDeclaration | InterfaceDeclaration = declaration.parent;
```
You can just declare this outside (as a `Node`) and rely on `isClassDeclaration` and `isInterfaceDeclaration` working as type predicates.

andy-ms commented on src/services/services.ts:
```diff
@@ -488,6 +502,63 @@ namespace ts {
+            const container: ClassDeclaration | InterfaceDeclaration = declaration.parent;
+            const baseTypeNode = getClassExtendsHeritageClauseElement(container);
+
+            if (baseTypeNode) {
```
Early return

andy-ms commented on src/services/services.ts:
```diff
@@ -488,6 +502,63 @@ namespace ts {
+
+        if (isClassDeclaration(declaration.parent) || isInterfaceDeclaration(declaration.parent)) {
+            const container: ClassDeclaration | InterfaceDeclaration = declaration.parent;
+            const baseTypeNode = getClassExtendsHeritageClauseElement(container);
```
There's a function `getBaseTypes` in `checker.ts` that might be better suited. You could expose it internally by adding `/* @internal */ getBaseTypes(): ReadonlyArray<BaseType>;` to `interface TypeChecker`.
The advantages are that this gets you both classes and interfaces, and you don't have to call `getTypeAtLocation`.

andy-ms commented on tests/cases/fourslash/jsDocInheritDoc.ts:
```diff
@@ -0,0 +1,49 @@
+////const p1 = b.property1/*4*/;
+////const p2 = b.property2/*5*/;
+
+verify.baselineQuickInfo();
```
This would be breifer using just `verify.quickInfoAt` -- `baselineQuickInfo` is more so we can check all the `displayParts`, but that's not what we're testing here.

andy-ms commented:
What if we just inherited documentation by default without requiring users to add `/** @inheritDoc */` to everything?

sjbarag commented:
Thanks for the review, @andy-ms!  I'll get this fixed up today or tomorrow.

> What if we just inherited documentation by default without requiring users to add `/** @inheritDoc */` to everything?

I think it'd be pretty intuitive to be honest.  I'd considered doing that here but didn't want to take too big of a step 😃  Should be relatively simple to get working, so I'll switch to that approach.

Thanks again! ❤️

sjbarag commented:
Got a few updates here based on your feedback!

1. Rebased on 9c96ee
2. Added early returns
3. Used `ts.getJSDocTags`
4. Used `verify.quickInfoAt` in tests

I unfortunately don't see a way to get `typeChecker.getBaseType` to work, since it accepts an `InterfaceType`.  Perhaps I'm missing something though?

Thanks for your patience while I updated this :smile:

andy-ms commented:
OK, I was wrong about `getBaseTypes` -- it will only return the `extends` type, not any interfaces.
I've managed to eliminate duplicate code between the `extends` and `implements` cases though:
```ts
function findInheritedJSDocComments(declaration: Declaration, propertyName: string, typeChecker: TypeChecker): SymbolDisplayPart[] {
    return flatMap(getAllSuperTypeNodes(declaration), superType => {
        const baseProperty = typeChecker.getPropertyOfType(typeChecker.getTypeAtLocation(superType), propertyName);
        return baseProperty && baseProperty.getDocumentationComment(typeChecker);
    });
}

function getAllSuperTypeNodes(declaration: Declaration): ReadonlyArray<TypeNode> {
    const container = declaration.parent;
    if (!container || (!isClassDeclaration(container) && !isInterfaceDeclaration(container))) {
        return emptyArray;
    }
    const extended = getClassExtendsHeritageClauseElement(container);
    const types = extended ? [extended] : emptyArray;
    return isClassLike(container) ? concatenate(types, getClassImplementsHeritageClauseElements(container)) : types;
}
```

andy-ms commented on src/services/types.ts:
```diff
@@ -32,7 +32,7 @@ namespace ts {
         getName(): string;
         getDeclarations(): Declaration[] | undefined;
-        getDocumentationComment(): SymbolDisplayPart[];
+        getDocumentationComment(typeChecker?: TypeChecker): SymbolDisplayPart[];
```
Based on https://github.com/Microsoft/TypeScript/pull/19455#discussion_r147214986 maybe it would be better to just take a compile-time breaking change and make this `typeChecker: TypeChecker | undefined`. @mhegazy

sjbarag commented on src/services/services.ts:
```diff
@@ -473,7 +488,16 @@ namespace ts {
+                if (this.declaration) {
+                    this.documentationComment = JsDoc.getJsDocCommentsFromDeclarations([this.declaration]);
+
+                    if (this.documentationComment.length === 0) {
```
Great point!  I agree.  This would make TypeScript's handling of `@inheritDoc` different than vanilla JSDoc's though, which ignores documentation on anything with `@inheritDoc`, e.g.:

```javascript
/** @class */
class Foo {
    /** Foo docs. */
    test() {
    }
}

/**
 * @class
 * @extends Foo
 */
class Bar {
    /**
     * Bar docs.
     * @inheritDoc
     */
    test() {
    }
}
```

results in JSDoc for `Bar#test` of "Foo docs" only.

That said, I think it's helpful for TypeScript users and I'd love to have it.  As long as you're okay with differing from upstream here, I'm 100% fine with it 👍

andy-ms replied:
You're right, http://usejsdoc.org/tags-inheritdoc.html says:
> The @inheritdoc tag indicates that a symbol should inherit its documentation from its parent class. Any other tags that you include in the JSDoc comment will be ignored.
> This tag is provided for compatibility with Closure Compiler. By default, if you do not add a JSDoc comment to a symbol, the symbol will inherit documentation from its parent.

This makes it a rather strange tag -- one whose sole purpose is to make all documentation on the current node ignored! Presumably people who don't want to see that comment in quick info should just make it a regular comment and not a doc comment. I think it would be a more useful tag if it meant to *add* docs from a parent in addition to anything on the child.

sjbarag replied:
Makes perfect sense to me.  I've updated the PR to support that :)

sjbarag commented:
Sorry! 424606c failed to build because I forgot to re-accept the baselines after switching from `typeChecker?: TypeChecker` to `typeChecker: TypeChecker | undefined`.  a839a80 should build 🤞

sjbarag commented:
Well this is embarrassing, I forgot to update `APISample_jsdoc.ts` to match the interface change.  Should be fixed now (for real (really (I mean it 😅)))

sjbarag commented:
Rebased on 1a7a587a9e!  Dang, `master` moves _quick_ 😄

andy-ms commented:
Thanks!

cliffkoh commented:
Thank you @sjbarag, this is super awesome!
